src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -80,11 +84,60 @@
                 out.add(new Tuple2<>(pair.key(header), pair));
             }
             return out;
-        }).groupByKey();
+        }).groupByKey(numReducers);
 
         return markPairedEnds(keyedPairs, scoringStrategy, finder, header);
     }
 
+    /**
+     * Like <code>groupByKey</code>, but assumes that values are already sorted by key, so no shuffle is needed,
+     * which is much faster.
+     * @param rdd the input RDD
+     * @param <K> type of keys
+     * @param <V> type of values
+     * @return an RDD where each the values for each key are grouped into an iterable collection of
+     */
+    static <K, V> JavaPairRDD<K, Iterable<V>> spanByKey(JavaPairRDD<K, V> rdd) {
+        return rdd.mapPartitionsToPair(iter -> () -> spanningIterator(iter));
+    }
+
+    /**
+     * An iterator that groups values having the same key into iterable collections.
+     * @param iterator an iterator over key-value pairs
+     * @param <K> type of keys
+     * @param <V> type of values
+     * @return an iterator over pairs of keys and grouped values
+     */
+    static <K, V> Iterator<Tuple2<K, Iterable<V>>> spanningIterator(Iterator<Tuple2<K, V>> iterator) {
+        final PeekingIterator<Tuple2<K, V>> iter = Iterators.peekingIterator(iterator);
+        return new AbstractIterator<Tuple2<K, Iterable<V>>>() {
+            @Override
+            protected Tuple2<K, Iterable<V>> computeNext() {
+                K key = null;
+                List<V> group = Lists.newArrayList();
+                while (iter.hasNext()) {
+                    if (key == null) {
+                        Tuple2<K, V> next = iter.next();
+                        key = next._1();
+                        V value = next._2();
+                        group.add(value);
+                        continue;
+                    }
+                    K nextKey = iter.peek()._1(); // don't advance...
+                    if (nextKey.equals(key)) {
+                        group.add(iter.next()._2()); // .. unless the keys matchMark
+                    } else {
+                        return new Tuple2<>(key, group);
```

Comment:
I think you have to keep accumulating reads until you reach a new name (not a new key) -- then break into groups according to key. Otherwise you'll create, eg.,  4 groupings given a sequence of keys like `RG1|Foo`, `RG2|Foo`, `RG1|Foo`, `RG2|Foo`, even though there are only 2 keys.

We only know that the reads are sorted by name -- we can't assume any secondary sorting by read group (the other component of the keys).